### DISCUSSION TITLE
Is using :not(:not(+~>)) more efficient than using :has(+~>)?

### DISCUSSION THREAD

**Discussion by u-RraaLL** [uBlockOrigin/uBlock-issues#2109] [Q&A] [answered]
Been wondering about it occasionally. Shouldn't using double `:not()` for siblings and children be more efficient than using a procedural filter? At least until `:has()` bacomes a native selector in the browsers as well?

**Comment by gothic-bum** [comment, marked as answer]
https://developer.mozilla.org/en-US/docs/Web/CSS/:not#description

> The `:not` pseudo-class may not be nested, which means that `:not(:not(...))` is invalid.  

https://github.com/gorhill/uBlock/wiki/Procedural-cosmetic-filters#subjectnotarg

> Note that if _arg_ is valid CSS selector, uBO will not consider the `:not` operator to be a procedural one, it will rather consider the operator as being part of a CSS selector.

**Comment by gothic-bum** [comment]
@u-RraaLL

Update:

The use of `:not(:not(...))` is specifically mentioned as being invalid with [W3C Selectors Level 3](https://www.w3.org/TR/selectors-3/#negation):
> Negations may not be nested; `:not(:not(...))` is invalid.

However, it isn't mentioned with [W3C Selectors Level 4](https://www.w3.org/TR/selectors-4/#negation), and per https://github.com/mdn/content/issues/17313:
> The [current Level 4 spec](https://drafts.csswg.org/selectors/#negation) removes any mention of this limitation.

But the [MDN page](https://developer.mozilla.org/en-US/docs/Web/CSS/:not) for `:not()` still says:  

> The `:not()` pseudo-class requires a comma-separated list of one or more selectors as its argument. The list must not contain another negation selector or a pseudo-element.

**Reply by u-RraaLL** [reply]
Thanks, but this doesn't really matter. The `:not(:not(...))` idea will be moot once we have `:has(...)` as a css selector.

... Maybe it could act as a non-forgiving version of `:is(...)`, but I'm not sure.
